lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
    */
   auto update_criteria = get_default_update_criteria();
   session_state.set_fsm_state(CREATED, update_criteria);
-  MLOG(MDEBUG) << "State of session changed to "
-               << session_fsm_state_to_str(session_state.get_state());
+  uint32_t cur_version = session_state.get_current_version();
+  session_state.set_current_version(cur_version++, update_criteria);
+  MLOG(MINFO) << "State of session changed to "
+               << session_fsm_state_to_str(session_state.get_state())
+	       << " Current version is " << session_state.get_current_version();
+
   /* Update the m5gsm_cause and prepare for respone along with actual cause*/
   prepare_response_to_access(
       imsi, session_state, magma::lte::M5GSMCause::OPERATION_SUCCESS);
+
+  SessionState::SessionInfo sess_info;
+  sess_info.imsi       = imsi;
+  sess_info.ip_addr    = ip_addr;
+  sess_info.Pdr_rules_ = session_state.get_all_pdr_rules();
+  sess_info.Far_rules_ = session_state.get_all_far_rules();
+  session_state.sess_infocopy(&sess_info);
   pipelined_client_->set_upf_session(sess_info, call_back_void_upf);
   return;
 }
 
+/* Function to initiate release of the session in enforcer requested by AMF
+ * Go over session map and find the respective session of imsi
+ * Go over SessionState vector and dfnd the respective dnn (apn) 
+ * start terminating session process
+ */
+bool SessionStateEnforcer::m5g_release_session(
+      SessionMap& session_map, const std::string& imsi,
+      const std::string& dnn, SessionUpdate& session_update)
+{
+   bool exist = false;
+   auto map_element = session_map.find(imsi);
+   if((map_element == session_map.end())) {
+      MLOG(MERROR) << "No session found in SessionMap for IMSI " << imsi
+                   << " during releasing the session";
+      return false;
+   }
+
+   for (auto& session : map_element->second) {
+     auto config     = session->get_config();
+     auto session_id = session->get_session_id();
+     if (config.common_context.apn() == dnn){
+	/*Irrespective of any State of Session, release and terminate*/
+        SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
+        MLOG(MINFO) 
+		<< "Start terminating the release request for IMSI "
+                << imsi << " having session id " << session_id
+		<< " current state of session is "
+                << session_fsm_state_to_str(session->get_state());
+	exist = true;

```

Comment:
as recommended by @themarwhal if you use `find_session` she created. It is very useful and you will save lots of code and variables like this

```
    SessionSearchCriteria criteria(imsi, IMSI_AND_APN, session_id);
    auto session_it = find_session(session_map, criteria);
    if (!session_it) {
      MLOG(MERROR)
          << session_id
          << " not found when m5g_release_session";
      return False;
    }

	auto& session = **session_it;
	auto session_id = session->get_session_id()

```

Reply:
Adapted with recommended function on possible places